### PR TITLE
Restore #45447

### PR DESCRIPTION
**Pull request by RyanSkonnord**
Undo revert of https://github.com/getsentry/sentry/pull/45447 with an additional fix.

### CODE REVIEW

**Comment by RyanSkonnord** [comment]
The content of https://github.com/getsentry/sentry/pull/46488/commits/419eb6bbdca2a2a49a7ad0121531301c171d4cb2 was already reviewed in https://github.com/getsentry/sentry/pull/45447. New changes are in https://github.com/getsentry/sentry/pull/46488/commits/e3da1e79ffbd5f8c8753339730f994d6e22d9b22 (and subsequent commits, if any).

Should prevent a recurrence of [SENTRY-104N](https://sentry.sentry.io/issues/4044472416/).

**Review by corps — APPROVED** [review]

File: src/sentry/services/hybrid_cloud/__init__.py [hunk 1]
```diff
@@ -33,22 +38,137 @@
...
         pass
 
 
+def report_pydantic_type_validation_error(
+    field: pydantic.fields.ModelField, value: Any, errors: pydantic.error_wrappers.ErrorList
+) -> None:
+    with sentry_sdk.push_scope() as scope:
+        scope.set_level("warning")
+        scope.set_tag("field", field.name)
+        scope.set_extra("value_type", str(type(value)))
```
Comment (dashed) [inline src/sentry/services/hybrid_cloud/__init__.py:69]: @RyanSkonnord `set_extra()` isn't preferred anymore. `scope.set_context("pydantic", extra_payload)` is preferred: https://docs.sentry.io/platforms/python/enriching-events/context/#structured-context

File: src/sentry/services/hybrid_cloud/__init__.py [hunk 2]
```diff
@@ -57,9 +60,62 @@ def close(self) -> None:
...
+
+
+_hack_pydantic_type_validation()
+
+
 class RpcModel(pydantic.BaseModel):
     """A serializable object that may be part of an RPC schema."""
 
+    class Config:
+        arbitrary_types_allowed = True
```
Comment (dashed) [inline src/sentry/services/hybrid_cloud/__init__.py:117]: Do we still need this?

**Review by markstory — APPROVED** [review]

File: src/sentry/services/hybrid_cloud/__init__.py (hunk 2 above)
Comment (RyanSkonnord) [reply src/sentry/services/hybrid_cloud/__init__.py]: We don't; good catch.

**Pull request merged**